File: 4_Iris_e_Dropout/no.js
// no.js

let tf;
try {
    tf = require('@tensorflow/tfjs');
} catch (error) {
    console.warn("no.js: Pacote '@tensorflow/tfjs' não encontrado. Usando mock.");
    // Mock para compatibilidade
    tf = {
        layers: {
            dense: (config) => ({ apply: (t) => t, getConfig: () => config }),
            dropout: (config) => ({ apply: (t) => t, getConfig: () => config }),
            activation: (config) => ({ apply: (t) => t, getConfig: () => config }),
            add: () => ({ apply: (tensors) => tensors[0] }),
        },
        tidy: (fn) => fn(),
    };
}

/**
 * Representa um nó de processamento.
 */
class No {
    id;
    layers;

    /**
     * Cria uma instância de Nó.
     * @param {string} id - Identificador único do nó.
     * @param {object[]} layersConfig - Um array de configurações de camada.
     * Ex: [{type: 'dense', config: {units: 32}}, {type: 'dropout', config: {rate: 0.5}}]
     */
    constructor(id, layersConfig) {
        if (!id || !Array.isArray(layersConfig) || layersConfig.length === 0) {
            throw new Error("Nó: 'id' e um array 'layersConfig' são obrigatórios.");
        }
        this.id = id;
        this.layers = [];

        // Cria as camadas do TensorFlow a partir da configuração
        for (const layerInfo of layersConfig) {
            if (tf.layers[layerInfo.type]) {
                this.layers.push(tf.layers[layerInfo.type](layerInfo.config));
            } else {
                throw new Error(`Tipo de camada desconhecido no Nó '${id}': ${layerInfo.type}`);
            }
        }
    }

    /**
     * O "forward pass" para o nó.
     * @param {tf.Tensor[]} listaTensoresEntrada - Um array de tensores.
     * @returns {tf.Tensor} O tensor de saída do nó.
     */
    call(listaTensoresEntrada) {
        if (!Array.isArray(listaTensoresEntrada) || listaTensoresEntrada.length === 0) {
            throw new Error(`Nó '${this.id}': A entrada deve ser um array de tensores não vazio.`);
        }

        return tf.tidy(() => {
            // Primeiro, agrega as entradas se houver mais de uma
            let tensorAtual = listaTensoresEntrada.length === 1
                ? listaTensoresEntrada[0]
                : tf.layers.add().apply(listaTensoresEntrada);

            // Aplica cada camada em sequência
            for (const layer of this.layers) {
                tensorAtual = layer.apply(tensorAtual);
            }

            return tensorAtual;
        });
    }
}

module.exports = No;

File: 6_GAN_Text-to-Image/no.py
# no.py
import tensorflow as tf
from tensorflow.keras import layers

class No(layers.Layer):
    """
    Representa um nó de processamento customizável na arquitetura MyceliumAI.
    Funciona como um container flexível que pode abrigar múltiplas camadas do Keras,
    processando tensores de entrada em sequência e agregando múltiplas entradas
    através de uma camada Add interna.
    """
    def __init__(self, name, layers_config, **kwargs):
        super(No, self).__init__(name=name, **kwargs)
        self.layers_config = layers_config
        self.internal_layers = []
        
        # A camada Add é criada uma única vez no construtor para garantir a estabilidade do grafo.
        self.add_layer = layers.Add(name=f"{self.name}_add")

        # Constrói as camadas internas a partir da "receita" (layers_config)
        for i, layer_info in enumerate(self.layers_config):
            if hasattr(layers, layer_info['type']):
                # Gera um nome único e seguro para cada camada interna para evitar conflitos
                layer_name = f"{name}_internal_{i}_{layer_info['type']}"
                self.internal_layers.append(
                    getattr(layers, layer_info['type'])(name=layer_name, **layer_info['config'])
                )
            else:
                raise ValueError(f"Tipo de camada desconhecido no Nó '{name}': {layer_info['type']}")

    def call(self, inputs):
        """
        Define o "forward pass" do Nó.
        """
        tensor_atual = inputs
        
        # Se a entrada for uma lista de tensores, primeiro os agrega com a camada Add.
        if isinstance(inputs, list) and len(inputs) > 1:
            tensor_atual = self.add_layer(inputs)
        
        # Aplica as camadas internas em sequência no tensor.
        for layer in self.internal_layers:
            tensor_atual = layer(tensor_atual)
            
        return tensor_atual

    # O método get_config é essencial para que o Keras saiba como salvar e carregar o modelo.
    def get_config(self):
        config = super(No, self).get_config()
        config.update({"layers_config": self.layers_config})
        return config